---
title: "scRNA-Seq Analysis with COVID-19 Data"
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
author: "Arjun Subramanian"
date: "10/17/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error=FALSE, warning=FALSE, message=FALSE, results = "hide")
```

$$\\[0in]$$

# Project Overview

In this project, I aim to provide an overview of data analysis associated with **scRNA-Seq (single-cell RNA sequencing)**, by creating cell clustering maps.  

### scRNA-Seq

scRNA-Seq measures the activity of thousands of genes within single cells. It practically accomplishes this by measuring the transcriptome (the sum of mRNA or messenger RNA transcripts) of many single cells, usually from one tissue.

The main goal of scRNA-Seq is to ["characterize heterogeneity across cells"](https://osca.bioconductor.org/feature-selection.html) by measuring gene expression differences between single cells. 

By allowing us to cell-specific expression differences, scRNA-Seq gives us detailed information that is useful for a number of applications. For example, we can identify gene expression patterns between individuals who have and do not have a certain disease.

### The Dataset

This project uses raw data from the [following paper](https://www.nature.com/articles/s41591-020-0901-9), titled "Single-cell landscape of bronchoalveolar immune cells in patients with COVID-19." 

This paper was written by Mingfeng Liao, Yang Liu, Jing Yuan, and several others from the Institute of Hepatology in the National Clinical Research Center for Infectious Disease at Shenzhen Third People's Hospital. It was published on May 12, 2020.

The focus of this paper was identifying respiratory immune characteristics associated with severe acute respiratory syndrome (SARS)-CoV2, through scRNA-Seq analysis of immune cells from broncheoalveolar lavage fluid. This fluid was collected from patients with varying severity of COVID-19 as well as healthy controls.

**Bronchoalveolar lavage fluid (BALF)** is fluid collected from a lung segment, through a diagnostic technique called bronchoscopy. In bronchoscopy, fluid is squirted into a small part of the lung and is promptly recollected for analysis. After it is collected, BALF mainly contains alveoli secretions (where gas exchange takes place). A typical application for BALF is diagnosing lung disease.

As COVID-19 is a respiratory disease, performing scRNA-Seq on cells from BALF can yield insights into how this disease affects gene expression of our cells. This paper explicitly states that it examines immune cells. Immune cells have multiple subtypes including macrophages, T cells, and B cells.

**Note: ** This paper has raw data from 4 controls and 9 people with COVID-19. For the purposes of this project, I decided to only analyze 3 controls because performing cell alignment on the fourth control (with data from a separate source) caused Excel to crash multiple times. 

$$\\[0in]$$

# Following Along

If you want to follow along, go to the [GitHub repository for this project](https://github.com/arjuncsubra/scRNA-Seq-with-COVID-19-Data). Make sure to download the repository and unzip it - it contains a folder called "scRNA-Seq-with-COVID-19-Data-master".
After that, do the following:

1. [Download and install R](https://cloud.r-project.org/). Make sure that the version is the latest R version (4.0.2 as of August 2020).
2. [Download and install RStudio](https://rstudio.com/products/rstudio/download/).
3. [Download and install 7Zip](https://www.7-zip.org/download.html) as it will easily allow to handle .tar.gz files. If you are on macOS, [Keka](https://www.keka.io/en/) is a great alternative.

Once you have installed R and RStudio, go to the GitHub respository download folder, "scRNA-Seq-with-COVID-19-Data-master". 

There, open the R Project file "scrna-seq-covid-project". Create a new R Script/R Markdown document (name it "my-r-file") and save it in "scRNA-Seq-with-COVID-19-Data-master". This is roughly what your main folder should look like:

![](C:/Users/arjun/Documents/checkpoint.PNG)

In R (specifically, in "my-r-file"), install [install Bioconductor](https://www.bioconductor.org/install/) and [devtools](https://www.r-project.org/nosvn/pandoc/devtools.html).

### Packages

**Make sure to install all of the following packages before starting Step 1** (you can do this by typing `install.packages("<package name>")` and `BiocManager::install("<package name>")`into the R console for CRAN packages and Bioconductor packages, respectively):

* **Seurat** (through CRAN)
* **SingleCellExperiment** (through Bioconductor)
* **scater** (through Bioconductor)
* **scran** (through Bioconductor)
* **cluster** (through CRAN)
* **PCAtools** (through Bioconductor)
* **hdf5r** (through CRAN) – if prompted, make sure to say "no" to installing from sources

Make sure to say "yes" to updates and additional software that is being downloaded along with these packages.

**Call these packages:**

```{r}
library(Seurat)
library(SingleCellExperiment)
library(scater)
library(scran)
library(cluster)
library(PCAtools)
library(hdf5r)
```

$$\\[0in]$$

# Step 1: Preprocessing

### Data Loading

The data we will be working with is an intermediate product of scRNA-Seq – it has already through the initial data processing stages (genome alignment). However, what is most important to know is that our data will contain genes as the rows and cells (from various samples) as the columns.

Our data is listed in NCBI's Gene Expression Omnibus (GEO) under the accession code [GSE145926](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE145926).

Once you have headed to the site, download the supplementary file (at the bottom of the page), "GSE145926_RAW.tar". 

Extract the .tar file and go to the folder with its contents. There, you will find many .h5 and .gz files.

Separate the 12 .h5 files from the 9 .gz files. Put the .h5 files in a separate folder (you can name this folder "count-matrices"): we do not need the .gz files for this project. Place the folder with the .h5 files in "scRNA-Seq-with-COVID-19-Data-master". This is what your "count-matrices" folder should look like:

![](C:/Users/arjun/Documents/countmatrices.PNG)

Go back to "my-r-file" (in RStudio) and set your working directory to "count-matrices". You can do this by clicking "Set Working Directory" under the Session tab.

Once that is done, enter the following commands:

```{r}
c51 <- Read10X_h5("GSM4475048_C51_filtered_feature_bc_matrix.h5")
c52 <- Read10X_h5("GSM4475049_C52_filtered_feature_bc_matrix.h5")
c100 <- Read10X_h5("GSM4475050_C100_filtered_feature_bc_matrix.h5")
c141 <- Read10X_h5("GSM4339769_C141_filtered_feature_bc_matrix.h5")
c142 <- Read10X_h5("GSM4339770_C142_filtered_feature_bc_matrix.h5")
c143 <- Read10X_h5("GSM4339771_C143_filtered_feature_bc_matrix.h5")
c144 <- Read10X_h5("GSM4339772_C144_filtered_feature_bc_matrix.h5")
c145 <- Read10X_h5("GSM4339773_C145_filtered_feature_bc_matrix.h5")
c146 <- Read10X_h5("GSM4339774_C146_filtered_feature_bc_matrix.h5")
c148 <- Read10X_h5("GSM4475051_C148_filtered_feature_bc_matrix.h5")
c149 <- Read10X_h5("GSM4475052_C149_filtered_feature_bc_matrix.h5")
c152 <- Read10X_h5("GSM4475053_C152_filtered_feature_bc_matrix.h5")
```

### Cell Alignment

There is one problem with this data: **many cells do not have complementary metadata** (in this case, cell metadata refers to cell-specific attributes, including the cell type and which person/group the cell came from). With no or incomplete metadata, it is much harder to generate meaningful analyses later onwards.


