---
title: "scRNA-Seq Analysis with COVID-19 Data"
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
author: "Arjun Subramanian"
date: "10/17/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error=FALSE, warning=FALSE, message=FALSE, results = "hide")
```

$$\\[0in]$$

# Project Overview

In this project, I aim to provide an overview of data analysis associated with **scRNA-Seq (single-cell RNA sequencing)**, by creating cell clustering maps.  

### scRNA-Seq

scRNA-Seq measures the activity of thousands of genes within single cells. It practically accomplishes this by measuring the transcriptome (the sum of mRNA or messenger RNA transcripts) of many single cells, usually from one tissue.

The main goal of scRNA-Seq is to ["characterize heterogeneity across cells"](https://osca.bioconductor.org/feature-selection.html) by measuring gene expression differences between single cells. 

By allowing us to cell-specific expression differences, scRNA-Seq gives us detailed information that is useful for a number of applications. For example, we can identify gene expression patterns between individuals who have and do not have a certain disease.

### The Dataset

This project uses raw data from the [following paper](https://www.nature.com/articles/s41591-020-0901-9), titled "Single-cell landscape of bronchoalveolar immune cells in patients with COVID-19." 

This paper was written by Mingfeng Liao, Yang Liu, Jing Yuan, and several others from the Institute of Hepatology in the National Clinical Research Center for Infectious Disease at Shenzhen Third People's Hospital. It was published on May 12, 2020.

The focus of this paper was identifying respiratory immune characteristics associated with severe acute respiratory syndrome (SARS)-CoV2, through scRNA-Seq analysis of immune cells from broncheoalveolar lavage fluid. This fluid was collected from patients with varying severity of COVID-19 as well as healthy controls.

**Bronchoalveolar lavage fluid (BALF)** is fluid collected from a lung segment, through a diagnostic technique called bronchoscopy. In bronchoscopy, fluid is squirted into a small part of the lung and is promptly recollected for analysis. After it is collected, BALF mainly contains alveoli secretions (where gas exchange takes place). A typical application for BALF is diagnosing lung disease.

As COVID-19 is a respiratory disease, performing scRNA-Seq on cells from BALF can yield insights into how this disease affects gene expression of our cells. This paper explicitly states that it examines immune cells. Immune cells have multiple subtypes including macrophages, T cells, and B cells.

**Note: ** This paper has raw data from 4 controls and 9 people with COVID-19. For the purposes of this project, I decided to only analyze 3 controls because performing cell alignment on the fourth control (with data from a separate source) caused Excel to crash multiple times. 

$$\\[0in]$$

# Following Along

If you want to follow along, go to the [GitHub repository for this project](https://github.com/arjuncsubra/scRNA-Seq-with-COVID-19-Data). Make sure to download the repository and unzip it - it contains a folder called "scRNA-Seq-with-COVID-19-Data-master".
After that, do the following:

1. [Download and install R](https://cloud.r-project.org/). Make sure that the version is the latest R version (4.0.2 as of August 2020).
2. [Download and install RStudio](https://rstudio.com/products/rstudio/download/).
3. [Download and install 7Zip](https://www.7-zip.org/download.html) as it will easily allow to handle .tar.gz files. If you are on macOS, [Keka](https://www.keka.io/en/) is a great alternative.

Once you have installed R and RStudio, go to the GitHub respository download folder, "scRNA-Seq-with-COVID-19-Data-master". 

There, open the R Project file "scrna-seq-covid-project". Create a new R Script/R Markdown document (name it "my-r-file") and save it in "scRNA-Seq-with-COVID-19-Data-master". This is roughly what your main folder should look like:

![](C:/Users/arjun/Documents/checkpoint.PNG)

In R (specifically, in "my-r-file"), install [install Bioconductor](https://www.bioconductor.org/install/) and [devtools](https://www.r-project.org/nosvn/pandoc/devtools.html).

### Packages

**Make sure to install all of the following packages before starting Step 1** (you can do this by typing `install.packages("<package name>")` and `BiocManager::install("<package name>")`into the R console for CRAN packages and Bioconductor packages, respectively):

* **Seurat** (through CRAN)
* **SingleCellExperiment** (through Bioconductor)
* **scater** (through Bioconductor)
* **scran** (through Bioconductor)
* **cluster** (through CRAN)
* **PCAtools** (through Bioconductor)
* **hdf5r** (through CRAN) – if prompted, make sure to say "no" to installing from sources
* **AnnotationHub** (through Bioconductor)

Make sure to say "yes" to updates and additional software that is being downloaded along with these packages.

**Call these packages:**

```{r}
library(Seurat)
library(SingleCellExperiment)
library(scater)
library(scran)
library(cluster)
library(PCAtools)
library(hdf5r)
library(AnnotationHub)
```

$$\\[0in]$$

# Step 1: Preprocessing

### Data Loading

The data we will be working with is an intermediate product of scRNA-Seq – it has already through the initial data processing stages (genome alignment). However, what is most important to know is that our data will contain genes as the rows and cells (from various samples) as the columns.

Our data is listed in NCBI's Gene Expression Omnibus (GEO) under the accession code [GSE145926](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE145926).

Once you have headed to the site, download the supplementary file (at the bottom of the page), "GSE145926_RAW.tar". 

Extract the .tar file and go to the folder with its contents. There, you will find many .h5 and .gz files.

Separate the 12 .h5 files from the 9 .gz files. Put the .h5 files in a separate folder (you can name this folder "count-matrices"): we do not need the .gz files for this project. Place the folder with the .h5 files in "scRNA-Seq-with-COVID-19-Data-master". This is what your "count-matrices" folder should look like:

![](C:/Users/arjun/Documents/countmatrices.PNG)

Go back to "my-r-file" (in RStudio) and set your working directory to "count-matrices". You can do this by clicking "Set Working Directory" under the Session tab.

Once that is done, enter the following commands:

```{r}
c51 <- Read10X_h5("GSM4475048_C51_filtered_feature_bc_matrix.h5")
c52 <- Read10X_h5("GSM4475049_C52_filtered_feature_bc_matrix.h5")
c100 <- Read10X_h5("GSM4475050_C100_filtered_feature_bc_matrix.h5")
c141 <- Read10X_h5("GSM4339769_C141_filtered_feature_bc_matrix.h5")
c142 <- Read10X_h5("GSM4339770_C142_filtered_feature_bc_matrix.h5")
c143 <- Read10X_h5("GSM4339771_C143_filtered_feature_bc_matrix.h5")
c144 <- Read10X_h5("GSM4339772_C144_filtered_feature_bc_matrix.h5")
c145 <- Read10X_h5("GSM4339773_C145_filtered_feature_bc_matrix.h5")
c146 <- Read10X_h5("GSM4339774_C146_filtered_feature_bc_matrix.h5")
c148 <- Read10X_h5("GSM4475051_C148_filtered_feature_bc_matrix.h5")
c149 <- Read10X_h5("GSM4475052_C149_filtered_feature_bc_matrix.h5")
c152 <- Read10X_h5("GSM4475053_C152_filtered_feature_bc_matrix.h5")
```

### Cell Alignment

There is one problem with this data: many cells do not have complementary metadata (in this case, cell metadata refers to cell-specific attributes, including the cell type and which person/group the cell came from). With no or incomplete metadata, it is much harder to generate meaningful analyses later onwards.

To resolve this, cells without metadata must be removed.

**Set your working directory to the "cell-alignment" folder in "scRNA-Seq-with-COVID-19-Data-master"** (it contains 12 .csv files starting with "c100numbers.csv"). The following code removes the cells that do not have associated metadata in the individual c objects (c51, c52, etc.).

```{r, }
c51.keep <- read.csv("c51numbers.csv", header=FALSE)
c51.vect <- c()
for (i in c51.keep){c51.vect <- append(c51.vect, i)}
c51.rev <- c51[,-c51.vect] #c51

c52.keep <- read.csv("c52numbers.csv", header=FALSE)
c52.vect <- c()
for (i in c52.keep){c52.vect <- append(c52.vect, i)}
c52.rev <- c52[,-c52.vect] #c52

c100.keep <- read.csv("c100numbers.csv", header=FALSE)
c100.vect <- c()
for (i in c100.keep){c100.vect <- append(c100.vect, i)}
c100.rev <- c100[,-c100.vect] #c100

c141.keep <- read.csv("c141numbers.csv", header=FALSE)
c141.vect <- c()
for (i in c141.keep){c141.vect <- append(c141.vect, i)}
c141.rev <- c141[,-c141.vect] #c141

c142.keep <- read.csv("c142numbers.csv", header=FALSE)
c142.vect <- c()
for (i in c142.keep){c142.vect <- append(c142.vect, i)}
c142.rev <- c142[,-c142.vect] #c142

c144.keep <- read.csv("c144numbers.csv", header=FALSE)
c144.vect <- c()
for (i in c144.keep){c144.vect <- append(c144.vect, i)}
c144.rev <- c144[,c144.vect] #c144

c143.keep <- read.csv("c143numbers.csv", header=FALSE)
c143.vect <- c()
for (i in c143.keep){c143.vect <- append(c143.vect, i)}
c143.rev <- c143[,-c143.vect] #c143

c145.keep <- read.csv("c145numbers.csv", header=FALSE)
c145.vect <- c()
for (i in c145.keep){c145.vect <- append(c145.vect, i)}
c145.rev <- c145[,-c145.vect] #c145

c146.keep <- read.csv("c146numbers.csv", header=FALSE)
c146.vect <- c()
for (i in c146.keep){c146.vect <- append(c146.vect, i)}
c146.rev <- c146[,-c146.vect] #c146

c148.keep <- read.csv("c148numbers.csv", header=FALSE)
c148.vect <- c()
for (i in c148.keep){c148.vect <- append(c148.vect, i)}
c148.rev <- c148[,-c148.vect] #c148

c149.keep <- read.csv("c149numbers.csv", header=FALSE)
c149.vect <- c()
for (i in c149.keep){c149.vect <- append(c149.vect, i)}
c149.rev <- c149[,-c149.vect] #c149

c152.keep <- read.csv("c152numbers.csv", header=FALSE)
c152.vect <- c()
for (i in c152.keep){c152.vect <- append(c152.vect, i)}
c152.rev <- c152[,-c152.vect] #c152
```

### Count Matrix Construction

Now that we have streamlined the number of cells in each sample, we need to merge the samples into a large count matrix with rows as genes and columns as individual cells.

To do this, all of our samples must have the same number of rows and columns. We have to remove the last row from the following samples to make this possible (this has no effect on our analyses):

```{r}
c141.rev.2 <- c141.rev[-nrow(c141.rev),]
c142.rev.2 <- c142.rev[-nrow(c142.rev),]
c143.rev.2 <- c143.rev[-nrow(c143.rev),]
c144.rev.2 <- c144.rev[-nrow(c144.rev),]
c145.rev.2 <- c145.rev[-nrow(c145.rev),]
c146.rev.2 <- c146.rev[-nrow(c146.rev),]
c148.rev.2 <- c148.rev[-nrow(c148.rev),]
c149.rev.2 <- c149.rev[-nrow(c149.rev),]
c152.rev.2 <- c152.rev[-nrow(c152.rev),]
```

These samples can be merged into a large count matrix. **Make sure to add in c144 before c143 due to the order of the metadata:**

```{r}
dgcmatrix.combined <- cbind(c51.rev,c52.rev,c100.rev)
dgcmatrix.combined <- cbind(dgcmatrix.combined, c141.rev.2, c142.rev.2)
dgcmatrix.combined <- cbind(dgcmatrix.combined, c144.rev.2, c143.rev.2)
dgcmatrix.combined <- cbind(dgcmatrix.combined, c145.rev.2, c146.rev.2)
dgcmatrix.combined <- cbind(dgcmatrix.combined, c148.rev.2, c149.rev.2)
dgcmatrix.combined <- cbind(dgcmatrix.combined, c152.rev.2)
```

$$\\[0in]$$

# Step 2: Metadata to SingleCellExperiment

Now, we can create an intermediate [Seurat object](https://github.com/satijalab/seurat/wiki#:~:text=Object%20Overview,manipulation%20of%20single%2Dcell%20data.&text=The%20Assay%20objects%20are%20designed,hashtags%2C%20or%20imputed%20gene%20values.), which allows us to easily add our metadata along with our count matrix. A Seurat object is essentially one type of container for storing single-cell data.

Before running the following code, **change your working directory back to "scRNA-Seq-with-COVID-19-Data-master", the folder you downloaded from GitHub.**

```{r}
metadata.seu <- read.csv("cell.metadata.csv", row.names=1)
seu <- CreateSeuratObject(counts=dgcmatrix.combined, assay="RNA", meta.data=metadata.seu) 
```

We can view the metadata to confirm that it has loaded properly:

```{r, results="markup",R.options=list(max.print=50)}
seu@meta.data
```

### Seurat to SingleCellExperiment

Another type of container for storing single-cell data is the SingleCellExperiment object, which we will be using to further analyze and explore our data. 

Fortunately, we can easily convert our Seurat object to a SingleCellExperiment object:


```{r}
sce <- as.SingleCellExperiment(seu)
```

We can view sce:

```{r, results="markup"}
sce
```

$$\\[0in]$$

# Step 3: Data Operations

We have created a SingleCellExperiment object, which is ready for further data processing.

### Quality Control

First, we will run **quality control (QC)** on our data. This ensures that we remove low-quality cells in our SingleCellExperiment object. If unaddressed, low-quality cells often distort the results we obtain from later steps in the analysis. For instance, such cells can form distinct clusters that do not reflect actual patterns in gene expression.

A common type of low quality cells is those with few expressed genes (in other words, scRNA-Seq failed to capture diversity in transcripts). Another common source of low-quality cells are those with high proportions of mitochondrial RNA. 

We can start quality control by identifying mitochondrial genes:

```{r}
location <- rowRanges(sce)
is.mito <- any(seqnames(location)=="MT")
```

We can now add QC statistics for each cell to the SingleCellExperiment object:

```{r}
df <- perCellQCMetrics(sce, subsets=list(Mito=is.mito))
df
```

Based on these statistics, we can identify what cells we will discard. We are discarding outlier cells (generally calculated by being more than 3 median absolute deviations from the median) when it comes to library sizes, expressed genes, and mitochondrial RNA.

```{r}
qc.lib <- isOutlier(df$sum, log=TRUE, type="lower")
qc.nexprs <- isOutlier(df$detected, log=TRUE, type="lower")
qc.mito <- isOutlier(df$subsets_Mito_percent, type="higher")

discard <- qc.lib | qc.nexprs | qc.mito
``` 

We can view a summary of these outliers:

```{r, results="markup"}
DataFrame(LibSize=sum(qc.lib), NExprs=sum(qc.nexprs), MitoProp=sum(qc.mito), Total=sum(discard))
```

Now, these outliers can be removed from our SingleCellExperiment:

```{r, results="markup"}
filtered <- sce[,!discard]
filtered
```
### Normalization and Feature Selection

We can now log-transform our counts. Log-transformation typically reduces skew in data (helping it conform better to a normal distribution):

```{r}
sce <- logNormCounts(sce)
```

Once we have normalized our data, we can move on to feature selection (gene selection). This is typically done by selecting highly variable genes. Selecting the most variable genes can help us better explore cell expression differences (as supposed to selecting all genes, some of which are cell-specific and tell us little about the whole dataset).

Before we select our highly variable genes (HVGs), we need to model gene variation:

```{r}
dec <- modelGeneVar(sce)
```

We can now select our top HVGs. I selected a numeric value (instead of a proportional value) because it allowed me to directly control the number of genes. This was because I wanted to make sure computation down the line was not too intensive.

```{r}
hvg <- getTopHVGs(dec, n=1000)
```


